solr/core/src/java/org/apache/solr/highlight/UnifiedSolrHighlighter.java

```diff
@@ -431,14 +435,27 @@ protected Locale parseLocale(String language, String country, String variant) {
 
     @Override
     protected Predicate<String> getFieldMatcher(String field) {
-      // TODO define hl.queryFieldPattern as a more advanced alternative to hl.requireFieldMatch.
 
       // note that the UH at Lucene level default to effectively "true"
       if (params.getFieldBool(field, HighlightParams.FIELD_MATCH, false)) {
         return field::equals; // requireFieldMatch
-      } else {
-        return NOT_REQUIRED_FIELD_MATCH_PREDICATE;
       }
+
+      String[] queryFieldPattern =
+          params.getFieldParams(field, HighlightParams.QUERY_FIELD_PATTERN);
+      if (queryFieldPattern != null && queryFieldPattern.length != 0) {
+
+        Collection<String> indexedFields = solrIndexSearcher.getDocFetcher().getIndexedFieldNames();
+
+        final Set<String> fields =
+            Set.of(expandWildcardsInHighlightFields(indexedFields, queryFieldPattern));
+        if (fields.isEmpty()) {
+          return NO_FIELD_MATCH_PREDICATE;
+        }
```

Comment:
These lines are fine but I just want to point out that it isn't necessary and it's a dubious optimization.

Reply:
Ah, yes, structural carry-over from the `if (predicate == null) {` that was there earlier but empty set would handle it implicitly, good spot.

Reply:
Fixed in https://github.com/apache/solr/pull/757/commits/0ac02d374fa95ac590c6e83d7b95642b1f813641 commit.